### ISSUE TITLE
[Help!] Dart/Flutter bindings

### ISSUE THREAD

**Issue by alfonsogarciacaro** [fable-compiler/Fable#2878]
I think we discussed it before but it'll be nice to have a semi-automatic way to generate Dart/Flutter bindings so users can start writing apps when the compiler is ready, so I'm opening this issue for discussion.

I've uploaded a dead-simple Elmish Flutter app here: https://github.com/alfonsogarciacaro/fable-flutterapp

I manually wrote a few bindings for the things I needed in [this file](https://github.com/alfonsogarciacaro/fable-flutterapp/blob/1a80389301c32f8a763bdac93f60c864ee07608d/src/Flutter.Material.fs). In general, bindings are very similar to the JS ones, there's some discussion about writing bindings in general for Snake Island in #2779. A couple of specifics for Dart:

- `NamedParams` attribute: used to indicate the method has named params in Dart. If the named params start from the 2nd or 3rd argument you can pass `fromIndex` (zero based) as in `[<NamedParams(fromIndex=1)>]`
- `IsConst` indicates the method (or constructor in the case of classes) should be called with `const` if all arguments are constants

Ideally the bindings should be extracted from [this repo](https://github.com/flutter/flutter/tree/master/packages/flutter/lib). In previous conversations @Nhowka commented there are introspection tools for Dart so maybe we could use those, or is it simpler to just write a simple custom parser? Pinging also @adelarsq who also has experience with Dart and @davedawkins because he's good at writing scripts for code generation as he did with the Feliz API 👼

**Comment by davedawkins** [comment]
Let me take a look then

**Comment by davedawkins** [comment]
New to flutter and dart. I have flutter sdk installed, extensions set up. This example app is building just fine .. I just can't see how to view it as an app?  Attempt to view web/index.html with LiveServer gives console errors. Any clues? Twitter DM to avoid noise here perhaps.

**Comment by davedawkins** [comment]
The Mirror reflection API looks promising.

**Comment by davedawkins** [comment]
Wow. F# app on macbook
<img width="806" alt="image" src="https://user-images.githubusercontent.com/285421/168040663-6e0b5e36-fb0f-4a05-878b-e4e05af5b8ca.png">

**Comment by davedawkins** [comment]
A very quick summary of my findings so far:

- Flutter is based upon dart. Flutter knows what dart is, dart doesn't know what flutter is
- 
- Flutter can create a MacOS app, so we can write output bindings direct to file system using reflection. However:

- Dart has Mirrors for reflection, but Flutter does not support Mirrors => we can't reflect on MaterialApp, Scaffold etc from a flutter app. An attempt to use Mirrors in Flutter produces a compilation/runtime error (I can't recall), and research on Stackoverflow backs this up

- Dart cannot (easily) load the flutter packages (eg 'package:flutter/material.dart', or even '../../flutter/lib/material.dart' eg) => this blocks the approach of using a Dart application to reflect on the flutter packages. Searches on "dart applications that use Flutter" result in "use flutter", understandably.

- Flutter has a relection package named "reflectable" which looks promising, but a simple example didn't compile. 

What to look at next:
- Challenge any of the "cannot" and "does not" findings above - I've tried and failed but it doesn't mean someone else can't make it work.
- Find an example of "reflectable" that compiles and works to understand it better
- Reference parser for Dart: https://github.com/dart-lang/sdk/wiki/The-Dart-specification-parser. We may be able to use this to parse Material.dart etc and generate the bindings
- Hand-made fuzzy parser. Once I understand the mapping that @alfonsogarciacaro  made between material.dart and Flutter.Material.fs I'll have a feel for how well this approach might work

**Comment by Nhowka** [comment]
[Dartdoc](https://github.com/dart-lang/dartdoc) is the tool that generates [this kind](https://api.dart.dev/stable/2.17.0/dart-collection/Queue-class.html) of documentation and the ones on pub.dev that already describes the public API for each library. Looking at the html it generates it could be feasible to parse them directly, but I still believe that the ideal solution could be reviving [this issue](https://github.com/dart-lang/dartdoc/issues/1667) so we could use the output to generate our bindings or forking/rewriting to our needs to generate them directly.

It even has an `--auto-include-dependencies` flag that generates documentation files for direct and indirect dependencies, so pretty powerful tool.

**Comment by adelarsq** [comment]
Nice! I will help on this 🤗

**Comment by Nhowka** [comment]
When importing the dartdoc package, we can create a `Dartdoc` instance calling the constructor `Dartdoc.withEmptyGenerator`. There is a brittle `generator` field on it with a `@visibleForTesting` annotation that we could leverage. The [`Generator`](https://github.com/dart-lang/dartdoc/blob/2549851dfba5bba2f25df903d37a409ce9f36fd4/lib/src/generator/generator.dart#L43) receives a `PackageGraph` that describes everything on the package and a `FileWriter` to write the result. 

I'm doing some experiments with it to learn the layout to extract everything we need, but that way we probably can use the tool as is.

**Comment by adelarsq** [comment]
A hand-made parser would make sense. So would be possible to extend [ts2fable](https://github.com/fable-compiler/ts2fable) to support Dart and more languages?

**Comment by Nhowka** [comment]
I upload the test file I made in [this repository](https://github.com/Nhowka/DartToFableBindings). I started handling only the constructors for now and got [this result for the flutter library](https://gist.github.com/Nhowka/4590ddb89d42cce93551697fff8009fd#file-flutterbindings-fs). Function types, the generics on the arguments and the translation from core types are still not correctly handled, but it's already a start.

While it's cool that it's using the metadata from the language and loading everything needed into the context so we could handle as we want, it makes it terribly slow. I ran it against my [elmish-like lib](https://pub.dev/packages/mvu_layer) with almost no dependencies and it takes around 2 minutes to reach the generator.

**Comment by davedawkins** [comment]
Amazing. The speed issue makes it hard to test/debug, but once completed this thing only needs to be run occasionally (to keep bindings in sync), I would think?

**Comment by Nhowka** [comment]
Yes, we could plug some pre-analysis step to check for changes only start the generation if there is a new or different version package. Maybe we could create a project and save a copy of the pubspec.lock and compare them. If there is a change, we regenerate all files again as it is easier than detecting changes on a single package.

For the named constructors I mapped them as static methods that return an instance of the class. Is that the right way to map them?

**Comment by alfonsogarciacaro** [comment]
Thanks everybody for your help! This is great @Nhowka, even if it requires some manual tweaking it's already a great help to make some demos/samples for the alpha release 🎉 As @davedawkins says speed probably it's not an issue if we only need to run the tool once in a while. For prototyping, is it faster if we work with a small code sample or most of the boot up time is taken by Dartdoc.

> @adelarsq About ts2fable, probably the only part we could reuse would be [printing](https://github.com/fable-compiler/ts2fable/blob/master/src/print.fs) and it has its own AST which may not have all the information we need (const constructors, named params) so it may be faster for now to do things in Dart directly as @Nhowka did 👍

**Comment by alfonsogarciacaro** [comment]
About compiling named constructors as static members, this is fine as in Dart they're syntactically the same in the call site. You can also use the `[<IsConst>]` attribute in a static member if needed: https://github.com/alfonsogarciacaro/fable-flutterapp/blob/1a80389301c32f8a763bdac93f60c864ee07608d/src/Flutter.Material.fs#L65

**Comment by alfonsogarciacaro** [comment]
@Nhowka I'm trying to edit your code a bit to get bindings for the constructors of the Material widgets. See this commit: https://github.com/alfonsogarciacaro/DartToFableBindings/commit/b7728ac5afbf9fd829585cd2765c9f0b14afb1e0

I tried with a test Dart package and it seemed to work fine, however I downloaded the [flutter repo](https://github.com/flutter/flutter), followed the [instructions here](https://github.com/flutter/flutter/wiki/Setting-up-the-Framework-development-environment) and managed to run `flutter analyze` in packages/flutter dir without issues. However when I try to run the fsgen from that dir I don't get any output. This is the command I'm using: 

```
dart run ../../../DartToFableBindings/bin/fsgen.dart \
  --exclude 'dart:async,dart:collection,dart:convert,dart:core,dart:developer,dart:io,dart:isolate,dart:math,dart:typed_data,dart,dart:ffi,dart:html,dart:js,dart:ui,dart:js_util' \
  --show-progress  \
  --no-auto-include-dependencies  \
  --no-validate-links  \
  --no-verbose-warnings  \
  --no-allow-non-local-warnings \
  --no-allow-tools
```

I only get many errors, this is the shortened log (only start and finish):

```
../../../../AppData/Local/Pub/Cache/hosted/pub.dartlang.org/dartdoc-5.0.1/lib/src/model/model_element.dart:706:14: Warning: Operand of null-aware operation '?.' has type 'LineInfo' which excludes null.
 - 'LineInfo' is from 'package:analyzer/source/line_info.dart' ('../../../../AppData/Local/Pub/Cache/hosted/pub.dartlang.org/analyzer-3.4.1/lib/source/line_info.dart').
      return lineInfo?.getLocation(nameOffset);
             ^
Documenting flutter...
  error: private API of package:Dart is reexported by libraries in other packages:
    from cupertino.Color: (file:///C:/Users/alfon/repos/flutter/bin/cache/pkg/sky_engine/lib/ui/painting.dart:94:7)
    referred to by cupertino: (file:///c:/users/alfon/repos/flutter/packages/flutter/lib/cupertino.dart:23:9)
    referred to by material: (file:///c:/users/alfon/repos/flutter/packages/flutter/lib/material.dart:21:9)
    referred to by painting: (file:///c:/users/alfon/repos/flutter/packages/flutter/lib/painting.dart:18:9)
    referred to by rendering: (file:///c:/users/alfon/repos/flutter/packages/flutter/lib/rendering.dart:22:9)
    referred to by widgets: (file:///c:/users/alfon/repos/flutter/packages/flutter/lib/widgets.dart:13:9)

...

  error: private API of package:Dart is reexported by libraries in other packages:
    from services.ByteData: (file:///C:/Users/alfon/repos/flutter/bin/cache/pkg/sky_engine/lib/typed_data/typed_data.dart:424:16)
    referred to by services: (file:///c:/users/alfon/repos/flutter/packages/flutter/lib/services.dart:11:9)
  error: private API of package:Dart is reexported by libraries in other packages:
    from services.ByteData.ByteData: (file:///C:/Users/alfon/repos/flutter/bin/cache/pkg/sky_engine/lib/typed_data/typed_data.dart:428:20)
    referred to by services: (file:///c:/users/alfon/repos/flutter/packages/flutter/lib/services.dart:11:9)
  error: private API of package:Dart is reexported by libraries in other packages:
    from services.ByteData.view: (file:///C:/Users/alfon/repos/flutter/bin/cache/pkg/sky_engine/lib/typed_data/typed_data.dart:459:20)
    referred to by services: (file:///c:/users/alfon/repos/flutter/packages/flutter/lib/services.dart:11:9)
  error: private API of package:Dart is reexported by libraries in other packages:
    from services.ByteData.sublistView: (file:///C:/Users/alfon/repos/flutter/bin/cache/pkg/sky_engine/lib/typed_data/typed_data.dart:481:20)
    referred to by services: (file:///c:/users/alfon/repos/flutter/packages/flutter/lib/services.dart:11:9)
-
dartdoc 5.0.1 (/C:/Users/alfon/repos/DartToFableBindings/bin/fsgen.dart) failed: Command executables must exist. The file "c:\users\alfon\repos\flutter\bin\cache\dart-sdk\bin\dart" does not exist for tool snippet..
```

Did you manage to run the fsgen from the flutter repo? Can you help? 🙏

**Comment by Nhowka** [comment]
Flutter is [somewhat unique](https://github.com/dart-lang/dartdoc/blob/a535f048b3235baba39a4bd2ffe82648ad208b54/example/README.md#flutter). They generate a dummy pubspec to make it compatible with dartdoc. Maybe you can get some output by creating a dummy empty flutter project that depends only on flutter and generating all bindings using it as a base.

**Comment by alfonsogarciacaro** [comment]
You're right, that worked! Now I've a ton of bindings... I only need to make them compile 😅 Getting excited now!

**Comment by bentok** [comment]
I'm gonna take a stab at getting that code generator to produce translations for function types. I'm using @alfonsogarciacaro's latest changes to the generator and it seems like functions are still not being translated. Correct me if I'm wrong on that, and I'm open to suggestions if anyone already has ideas.

**Comment by alfonsogarciacaro** [comment]
Thanks a lot for your help @bentok! I think my latest changes are in this branch: https://github.com/alfonsogarciacaro/DartToFableBindings/tree/material-widgets

I was making several attempts and also did some manual changes. I think this code can output the module functions, but in any case the generator is a bit dirty at the moment, so some cleanup would be much appreciated. See the comment from @Nhowka about the dummy project necessary to make the doc generator work 👍 https://github.com/alfonsogarciacaro/DartToFableBindings/blob/f001c9621d018a082a37cd593745d7549d33f72c/lib/fsgen.dart#L358-L361

**Comment by bentok** [comment]
> Flutter is [somewhat unique](https://github.com/dart-lang/dartdoc/blob/a535f048b3235baba39a4bd2ffe82648ad208b54/example/README.md#flutter). They generate a dummy pubspec to make it compatible with dartdoc. Maybe you can get some output by creating a dummy empty flutter project that depends only on flutter and generating all bindings using it as a base.

@Nhowka do you think generating bindings for 3rd party Flutter libraries would require the same approach? I have made some modifications to both your code and @alfonsogarciacaro's fork and am getting close, but can't quite get it right.

For example, if I generate bindings for flutter_secure_storage, it will generate dozens of files - some of which have only a few bindings and others that have more. However it doesn't seem like any of the files have _all_ of the bindings.

**Comment by Nhowka** [comment]
@bentok, I believe that for typical packages with standard pubspec files, the same surface outputted on their [dartdoc documentation](https://pub.dev/documentation/flutter_secure_storage/latest/flutter_secure_storage/flutter_secure_storage-library.html) on pub.dev should be available to us. Could there be some filter pruning some of those before they reach the code generation stage?

**Comment by kaashyapan** [comment]
Dart makes extensive use of [Code generation using annotations](https://medium.com/flutter-community/part-2-code-generation-in-dart-annotations-source-gen-and-build-runner-bbceee28697b).

Can someone point me to how I can add these to FSharp code which will transform as annotations on dart ? Trying to see if I can help with this issue.

**Comment by johnstorey** [comment]
Hey, l've come late to this thread, but @bentok and I have connected to start working on this again. I'm able to reproduce everything up to @alfonsogarciacaro  getting errors on his first try to run this. But I have not yet worked out what to do once the empty flutter project is set up to get the bindings. I know it's been more than a year, but if someone remember and can give the exact step that would speed up the work a bit.

**Comment by octaviordz** [comment]
Hi everyone, I got the source code from https://github.com/alfonsogarciacaro/fable-flutterapp and after updating fable it compiled and ran correctly, but the auto/hot-reload does not work. Is there something that can be done to make this work?

I'm using vscode under Windows 10.
Thank you.

**Comment by MangelMaxime** [comment]
Hello,

in the readme of the repo there is a mention that you have to trigger hot reload manually.

Did you give it a try ?

I never used Flutter so I don’t know if there is some requirements like in JavaScript for it to work.

**Comment by octaviordz** [comment]
Hi, I had not tried the "Ctr+F5" that works, I was just manually updating the .dart generated file to trigger the hot reload. But my question was more in regard to what is the missing/link/configuration the 'build.cmd' is running fable in watch mode so when you update the source code it creates the .dart file(s) and vscode does load the updated .dart file(s) and like I said I was manually updating the .dart file to manual trigger hot reload hence my question is why if vscode 'knows' the file has changed the hot reload is not triggered.

Thank you.

**Comment by MangelMaxime** [comment]
Hello @octaviordz,

I am not sure I understood everything because it was a really long sentence 😅

But, the first thing I was check is what is the behaviour of hot reload in a pure Dart project because if the behaviour is the same then there is nothing that can be improve.

If not, then someone with Dart/Flutter knowledge will need to have a look at it.

**Comment by octaviordz** [comment]
Hi @MangelMaxime,
Hot reload works in a pure Dart project as expected.  Thank you for your response.

**Comment by MangelMaxime** [comment]
@endeavour The main developer for Dart target is not active anymore.

If someone want to step in to contribute to it, we welcome contributors.

**Comment by MangelMaxime** [comment]
I am doing some clean up in the issue.

Dart development is not really active anymore, so I am closing this ticket. If there are traction from the community again we can re-open.
